File: core/LowestCommonAncestorOfABinaryTree/LowestCommonAncestorOfABinaryTree.java
package com.dbc;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class LowestCommonAncestorOfABinaryTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    private final Map<Integer, TreeNode> parents = new HashMap<>();

    public void dfs(TreeNode root) {
        if (root.left != null) {
            this.parents.put(root.left.val, root);
            dfs(root.left);
        }
        if (root.right != null) {
            this.parents.put(root.right.val, root);
            dfs(root.right);
        }
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        dfs(root);
        Set<TreeNode> set = new HashSet<>();
        while (p != null) {
            set.add(p);
            p = this.parents.get(p.val);
        }
        while (q != null) {
            if (set.contains(q)) return q;
            q = this.parents.get(q.val);
        }
        return null;
    }
}

File: core/LowestCommonAncestorOfABinaryTree/LowestCommonAncestorOfABinaryTree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def lowestCommonAncestor(root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
    parents, visited = {}, {}
    def dfs(node):
        if node.left:
            parents[node.left.val] = node
            dfs(node.left)
        if node.right:
            parents[node.right.val] = node
            dfs(node.right)

    dfs(root)
    while p:
        visited[p.val] = True
        p = parents.get(p.val)
    while q:
        if visited.get(q.val, False):
            return q
        q = parents.get(q.val)

    return None



